### DISCUSSION TITLE
expected &mut {unknown}, found &MysqlConnection when fetching users

### DISCUSSION THREAD

**Discussion by AaronCarton** [diesel-rs/diesel#3485] [Q&A] [answered]
### Versions

- **Rust:** 1.66.1
- **Diesel:** 2.0.0
- **Database:** mysql:5.7
- **Operating System** Ubuntu WSL on Windows 11

### Feature Flags

- **diesel:**  mysql, serde_json

## Problem Description
I'm trying to set up a simple CRUD with Diesel & MySQL, but the following code does not want to accept `MysqlConnection` as a valid argument, whereas it had no issue accepting `PgConnection`

```rust
pub fn establish_connection() -> MysqlConnection {
    dotenv().ok();

    let database_url = env::var("DATABASE_URL").expect("DATABASE_URL must be set");
    MysqlConnection::establish(&database_url)
        .expect(&format!("Error connecting to {}", database_url))
}
```
```rust
pub fn get_users() -> Vec<User> {
    use crate::schema::users::dsl::*;

    let connection = establish_connection();
    let results = users.load::<User>(&connection).unwrap();
                                    //^ error here

    results
}
```

### What are you trying to accomplish?
Return a list of users from a MySQL database

### What is the expected output?
a list of users

### What is the actual output?
```
expected &mut {unknown}, found &MysqlConnection
the trait bound `(diesel::sql_types::Integer, Text, diesel::sql_types::Integer, Text, diesel::sql_types::Integer): load_dsl::private::CompatibleType<User, Mysql>` is not satisfied
the following other types implement trait `load_dsl::private::CompatibleType<U, DB>`:
  (ST0, ST1)
  (ST0, ST1, ST2)
  (ST0, ST1, ST2, ST3)
  (ST0, ST1, ST2, ST3, ST4)
  (ST0, ST1, ST2, ST3, ST4, ST5)
  (ST0, ST1, ST2, ST3, ST4, ST5, ST6)
  (ST0, ST1, ST2, ST3, ST4, ST5, ST6, ST7)
  (ST0, ST1, ST2, ST3, ST4, ST5, ST6, ST7, ST8)
and 24 others
required for `table` to implement `LoadQuery<'_, _, User>`
```

### Are you seeing any additional errors?
on Diesel 1.4.4 it gives the following error instead
```
the trait bound `(i32, std::string::String, i32, std::option::Option<std::string::String>, std::option::Option<i32>, std::option::Option<SystemTime>): Queryable<(diesel::sql_types::Integer, Text, diesel::sql_types::Integer, Text, diesel::sql_types::Integer), _>` is not satisfied
the following other types implement trait `Queryable<ST, DB>`:
  <(A, B) as Queryable<(SA, SB), __DB>>
  <(A, B, C) as Queryable<(SA, SB, SC), __DB>>
  <(A, B, C, D) as Queryable<(SA, SB, SC, SD), __DB>>
  <(A, B, C, D, E) as Queryable<(SA, SB, SC, SD, SE), __DB>>
  <(A, B, C, D, E, F) as Queryable<(SA, SB, SC, SD, SE, SF), __DB>>
  <(A, B, C, D, E, F, G) as Queryable<(SA, SB, SC, SD, SE, SF, SG), __DB>>
  <(A, B, C, D, E, F, G, H) as Queryable<(SA, SB, SC, SD, SE, SF, SG, SH), __DB>>
  <(A, B, C, D, E, F, G, H, I) as Queryable<(SA, SB, SC, SD, SE, SF, SG, SH, SI), __DB>>
and 24 others
required for `table` to implement `LoadQuery<_, User>`
```
## Checklist

- [x] I have already looked over the [issue tracker](https://github.com/diesel-rs/diesel/issues)  and the [disussion forum](https://github.com/diesel-rs/diesel/discussions)  for similar possible closed issues.
 
- [x] This issue can be reproduced on Rust's stable channel. (Your issue will be
  closed if this is not the case)
- [x] This issue can be reproduced without requiring a third party crate

**Comment by weiznich** [comment, marked as answer]
I converted this to a discussion as our issue tracker is the wrong place for support questions.

Other than that: The error message already contains a quite strong hint what's wrong there:

```
expected &mut {unknown}, found &MysqlConnection
```

Which means that you need a mutable reference to `.load()` not a shared reference. That also [matches the documentation](https://docs.diesel.rs/2.0.x/diesel/prelude/trait.RunQueryDsl.html#method.load)

As for the second part of the error message: That indicates a type mismatch between what your query returns and what your return type expects. As you did not provide all the relevant code it's not possible to give you more details on that, other than you either need to change your select clause or the definition of the `User` type.

**Reply by AaronCarton** [reply]
Yep, you're right.

adding `&mut` solved the connection issue
```rust
pub fn get_users() -> Vec<User> {
    use crate::schema::users::dsl::*;
    let connection = &mut establish_connection();
    let results = users.load::<User>(connection).unwrap();

    results
}
```

As for the second part, turns out I had marked some fields in my User model as `Optional< >` where they weren't in the database schema 😅

Thanks 👍
